packages/reason-css-parser/ppx/EmitType.re

```diff
@@ -0,0 +1,301 @@
+    | Repeat(_)
+    | Repeat_by_comma(_, _)
+    | Zero_or_more
+    | One_or_more
+    | At_least_one =>
+      let params = [ptyp_constr(txt @@ Lident("list"), params)];
+      mk_branch(type_, is_constructor, params);
+    };
+
+  let create_value_parser = (type_name, value) => {
+    let terminal_op = (kind, multiplier) => {
+      let (type_, is_constructor, params) =
+        switch (kind) {
+        | Keyword(name) => (first_uppercase(name), false, [])
+        | Data_type(name) =>
+          let name = value_name_of_css(name);
+          let params = [ptyp_constr(txt @@ Lident(name), [])];
+          (first_uppercase(name), false, params);
+        | Property_type(name) =>
+          let name = property_value_name(name) |> value_name_of_css;
+          let params = [ptyp_constr(txt @@ Lident(name), [])];
+          (first_uppercase(name), false, params);
+        | _ => raise(Unsupported)
+        };
+      apply_modifier(multiplier, type_, is_constructor, params);
+    };
+
+    let combinator_op = (kind, values) => {
+      switch (kind) {
+      | Xor =>
+        List.map(
+          fun
+          | Terminal(kind, multiplier) => terminal_op(kind, multiplier)
+          | _ => raise(Unsupported),
+          values,
+        )
+      | _ => raise(Unsupported)
+      };
+    };
+
+    switch (value) {
+    | Terminal(kind, multiplier) =>
+      mk_typ(type_name) @@ [terminal_op(kind, multiplier)]
+    | Combinator(kind, values) =>
+      mk_typ(type_name) @@ combinator_op(kind, values)
+    // | Group(value, multiplier) => group_op(value, multiplier)
+    // | Function_call(name, value) => function_call(name, value)
+    | _ => abstract_type(type_name)
+    };
+  };
+};
+
+let extract_ppx_content = (exp: Parsetree.expression) => {
+  switch (exp.pexp_desc) {
+  | Pexp_extension((
+      _,
+      PStr([
+        {
+          pstr_desc:
+            Pstr_eval(
+              {
+                pexp_desc: Pexp_constant(Pconst_string(value, loc, _delim)),
+                _,
+              },
+              _attrs,
+            ),
+          _,
+        },
+      ]),
+    )) => (
+      value,
+      loc,
+    )
+  | _ => failwith("expected a ppx extension")
+  };
+};
+
+
+let standard_types = {
+
+  open Ppxlib.Ast_builder.Default;
+
+  let abstract_type = name => {
+      type_declaration(
+        ~loc=Location.none,
+        ~name={txt: name, loc: Location.none},
+        ~params=[],
+        ~cstrs=[],
+        ~private_=Public,
+        ~manifest=None,
+        ~kind=Ptype_abstract,
+      );
+  };
+
+  [
+    abstract_type("integer"),
+    abstract_type("number"),
+    abstract_type("length"),
+    abstract_type("angle"),
+    abstract_type("time"),
+    abstract_type("frequency"),
+    abstract_type("resolution"),
+    abstract_type("percentage"),
+    abstract_type("ident"),
+    abstract_type("custom_ident"),
+    abstract_type("any_value"),
+    // abstract_type("string"), already represented by OCaml string type
```

Comment:
This will need to be a real definition, right?

Reply:
I think so, added this to make the code compile, i think that the simplest solution is make the types using metaquot, e.g:

```reason
[%stri type angle = [
  | `Deg(float)
  | `Grad(float)
  | `Rad(float)
  | `Turn(float)
]]
```